keps/sig-api-machinery/1872-manifest-based-admission-webhooks/README.md

```diff
@@ -0,0 +1,382 @@
+
+- The registration of webhooks registered through this process should be
+  protected from alteration by API requests made by cluster users.
+
+- It should be possible to alter webhooks registered through this new process
+  without restarting the kube-apiserver.
+ 
+## Proposal
+
+In short, this proposal is about augmenting `AdmissionConfig`'s plugin
+`configuration` [specification](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers)
+to include a path to a configuration file containing one of the following:
+- `admissionregistration.k8s.io/v1.ValidatingWebhookConfigurationList` 
+- `admissionregistration.k8s.io/v1.MutatingWebhookConfigurationList` 
+- v1.List with items of admissionregistration.k8s.io/v1.ValidatingWebhookConfiguration 
+- v1.List  with items of admissionregistration.k8s.io/v1.MutatingWebhookConfiguration.
+This configuration is loaded as manifest based webhooks in api servers. These
+webhooks are called by validating and mutating admission plugins along with
+dynamically loaded webhooks for relevant admission requests. This also means
+that these webhooks are not API visible objects and hence cannot be modified by
+an API user.
+
+### Naming
+All webhooks in the manifest file need to have unique names. If a new webhook
+configuration API object with a same name as a webhook in the manifest is added,
+both webhook would be invoked. Essentially, webhooks in the manifest file will
+be treated as belonging to a different domain from the webhooks registered
+through the API.
+
+### Metrics and audit annotations
+Metrics and audit annotations for webhooks registered through the manifest file
+will be surfaced differently. This allows the administrator to unique monitor
+activity of these webhooks as they would be the only party able to take action
+against issues with these webhooks. 
+
+All admission metrics for webhooks will have an additional label `manifest_based`
+with a value `true` for webhooks registered via manifest.
+
+Audit annotations for mutation webhooks will include another field `manifestBased`
+with a value `true` for webhooks registered via manifest.
+
+```
+<<[UNRESOLVED sig-instrumentation ]>>
+get feedback on how to indicate this in metrics
+<<[/UNRESOLVED]>>
+```
+
+### New AdmissionConfig schema
+
+A `webhooksFile` field is added to `configuration` field of a plugin object in
+`AdmissionConfig`.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: ValidatingAdmissionWebhook
+  configuration:
+    ....
+    webhooksFile: "<path-to-manifest-file>"
+- name: MutatingAdmissionWebhook
+  configuration:
+    ....
+    webhooksFile: "<path-to-manifest-file>"
+```
+
+### Reconfiguring manifest file
+
+The manifest file is watched so that the webhook configuration can be
+dynamically changed by editing the contents of the file.
+
+```
+<<[UNRESOLVED]>>
+need to define behavior in the following cases:
```

Comment:
one advantage to running from a file on disk is that, "this configuration is used".  For certificates, you can externally observe when the change is picked up.

I'd like to have a mechanism for indicating when this change has taken affect.